### PR TITLE
Pass messages as smart pointers E2E: phase one

### PR DESCRIPTION
**Pull request by yuliasherman**
- Problem Overview
BFT messages are used as raw pointers in the concord-bft code, as a result, we have memory leaks. This code has to be improved by starting to use managed pointers instead. This PR changes the onMessage API to pass messages as smart pointers and have initial support for them in the PreProcessor module.

- Testing Done
Runs on a reserved system with interruptions + Maestro high-end run.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -545,34 +534,37 @@ void ReplicaImp::onMessage<ClientRequestMsg>(ClientRequestMsg *m) {
...
-                << KVLOG(m->clientProxyId(), m->getCid(), m->requestSeqNum()));
-  return onMessage<ClientRequestMsg>(m);
+                << KVLOG(msg->clientProxyId(), msg->getCid(), msg->requestSeqNum()));
+  return onMessage<ClientRequestMsg>(std::make_unique<ClientRequestMsg>(msg.release()));
 }
 
 template <>
-void ReplicaImp::onMessage<ReplicaAsksToLeaveViewMsg>(ReplicaAsksToLeaveViewMsg *m) {
+void ReplicaImp::onMessage<ReplicaAsksToLeaveViewMsg>(std::unique_ptr<ReplicaAsksToLeaveViewMsg> message) {
+  auto *msg = message.release();
```
Comment (toly-kournik) [inline bftengine/src/bftengine/ReplicaImp.cpp:550]: i'd delete this line

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -545,34 +534,37 @@ void ReplicaImp::onMessage<ClientRequestMsg>(ClientRequestMsg *m) {
...
   if (activeExecutions_ > 0) {
-    pushDeferredMessage(m);
+    pushDeferredMessage(msg);
```
Comment (toly-kournik) [inline bftengine/src/bftengine/ReplicaImp.cpp:552]: message.release()

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 3]
```diff
@@ -545,34 +534,37 @@ void ReplicaImp::onMessage<ClientRequestMsg>(ClientRequestMsg *m) {
...
   }
   MDC_PUT(MDC_SEQ_NUM_KEY, std::to_string(getCurrentView()));
-  if (m->viewNumber() == getCurrentView()) {
+  if (msg->viewNumber() == getCurrentView()) {
     LOG_INFO(VC_LOG,
-             "Received ReplicaAsksToLeaveViewMsg " << KVLOG(m->viewNumber(), m->senderId(), m->idOfGeneratedReplica()));
-    viewsManager->storeComplaint(std::unique_ptr<ReplicaAsksToLeaveViewMsg>(m));
+             "Received ReplicaAsksToLeaveViewMsg "
+                 << KVLOG(msg->viewNumber(), msg->senderId(), msg->idOfGeneratedReplica()));
+    viewsManager->storeComplaint(std::unique_ptr<ReplicaAsksToLeaveViewMsg>(msg));
```
Comment (toly-kournik) [inline bftengine/src/bftengine/ReplicaImp.cpp:560]: `viewsManager->storeComplaint(std::move(message))`

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 4]
```diff
@@ -1451,8 +1445,9 @@ void ReplicaImp::sendCommitPartial(const SeqNum s) {
 }
 
 template <>
-void ReplicaImp::onMessage<PartialCommitProofMsg>(PartialCommitProofMsg *msg) {
+void ReplicaImp::onMessage<PartialCommitProofMsg>(std::unique_ptr<PartialCommitProofMsg> message) {
   metric_received_partial_commit_proofs_++;
+  auto *msg = message.release();
```
Comment (toly-kournik) [inline bftengine/src/bftengine/ReplicaImp.cpp:1450]: same

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 1 above)
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:550]: For now, there is no further handling of unique message pointers in the code. That's why all of them get released in all onMessage functions. The addition of proper handling is one of the next stages.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -1484,7 +1479,8 @@ void ReplicaImp::onMessage<PartialCommitProofMsg>(PartialCommitProofMsg *msg) {
 }
 
 template <>
-void ReplicaImp::onMessage<FullCommitProofMsg>(FullCommitProofMsg *msg) {
+void ReplicaImp::onMessage<FullCommitProofMsg>(std::unique_ptr<FullCommitProofMsg> message) {
+  auto *msg = message.release();
```
Comment (toly-kournik) [inline bftengine/src/bftengine/ReplicaImp.cpp:1483]: same

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 2 above)
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:552]: Why?

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 6]
```diff
@@ -1063,7 +1055,8 @@ bool ReplicaImp::validatePreProcessedResults(const PrePrepareMsg *msg, const Vie
 }
 
 template <>
-void ReplicaImp::onMessage<PrePrepareMsg>(PrePrepareMsg *msg) {
+void ReplicaImp::onMessage<PrePrepareMsg>(std::unique_ptr<PrePrepareMsg> message) {
+  auto *msg = message.release();
```
[the remark this answers is not in the document]
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:1059]: Good catch! It apparently leaked before.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 7]
```diff
@@ -1300,8 +1293,9 @@ void ReplicaImp::tryToAskForMissingInfo() {
 }
 
 template <>
-void ReplicaImp::onMessage<StartSlowCommitMsg>(StartSlowCommitMsg *msg) {
+void ReplicaImp::onMessage<StartSlowCommitMsg>(std::unique_ptr<StartSlowCommitMsg> message) {
   metric_received_start_slow_commits_++;
+  auto *msg = message.release();
```
[the remark this answers is not in the document]
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:1298]: Same

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 4 above)
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:1450]: Same

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 5 above)
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:1483]: Same

**Review by toly-kournik — APPROVED** [review]
LGTM as a first phase for gradual switch to smart ptrs.
It's essential to check with heap debugger.

File: bftengine/src/bftengine/ReplicaImp.cpp (hunk 2 above)
Comment (yuliasherman) [reply bftengine/src/bftengine/ReplicaImp.cpp:552]: See related comments

File: bftengine/src/preprocessor/messages/ClientPreProcessRequestMsg.hpp [hunk 8]
```diff
@@ -37,6 +37,7 @@ class ClientPreProcessRequestMsg : public ClientRequestMsg {
 };
 
 typedef std::unique_ptr<ClientPreProcessRequestMsg> ClientPreProcessReqMsgUniquePtr;
+typedef std::shared_ptr<ClientPreProcessRequestMsg> ClientPreProcessReqMsgSharedPtr;
```
Comment (glevkovich) [inline bftengine/src/preprocessor/messages/ClientPreProcessRequestMsg.hpp:40]: nit:
why not type aliasing (more modern c++):
```
using ClientPreProcessReqMsgUniquePtr = std::unique_ptr<ClientPreProcessRequestMsg>;
using ClientPreProcessReqMsgSharedPtr = std::shared_ptr<ClientPreProcessRequestMsg>;
```

File: bftengine/src/preprocessor/messages/PreProcessReplyMsg.hpp [hunk 9]
```diff
@@ -112,6 +112,7 @@ class PreProcessReplyMsg : public MessageBase {
   inline static preprocessor::PreProcessorRecorder* preProcessorHistograms_ = nullptr;
 };
 
+typedef std::unique_ptr<PreProcessReplyMsg> PreProcessReplyMsgUniquePtr;
```
Comment (glevkovich) [inline bftengine/src/preprocessor/messages/PreProcessReplyMsg.hpp:115]: type aliasing, same for all below places.

**Review by glevkovich — APPROVED** [review]
Approved.
I made a few comments and then stopped commenting on the same topic: I'm not sure if you should have split the use of the unique ptr at the top of the function in all places. If you already have a unique ptr and the function is simple, you could have done the release directly into the next function's container  or another function, and remove out all the deletes.

Anyway, we can keep doing that on future tasks.

**Pull request merged**